src/local/terrain.rs

```diff
@@ -0,0 +1,58 @@
+use std::mem::MaybeUninit;
+
+use js_sys::Uint8Array;
+
+use crate::{constants::Terrain, objects::RoomTerrain};
+
+use super::{xy_to_terrain_index, RoomXY, ROOM_AREA};
+
+pub struct LocalRoomTerrain {
+    bits: Box<[u8; ROOM_AREA]>,
+}
+
+impl LocalRoomTerrain {
+    /// Gets the terrain at the specified position in this room.
+    pub fn get(&self, xy: RoomXY) -> Terrain {
+        // SAFETY: RoomXY is always a valid coordinate.
+        let byte = unsafe { self.bits.get_unchecked(xy_to_terrain_index(xy)) };
+        // not using Terrain::from_u8() because `0b11` value, wall+swamp, happens
+        // in commonly used server environments (notably the private server default
+        // map), and is special-cased in the engine code; we special-case it here
+        match byte {
+            0b00 => Terrain::Plain,
+            0b01 | 0b11 => Terrain::Wall,
+            0b10 => Terrain::Swamp,
+            _ => panic!("unexpected terrain"),
+        }
```

Comment:
If we masked the byte with `& 0b11`, we could cause the panic to be optimized out. The only user facing change is that masking produces incorrect results if the terrain data ever updates (though we have so many other places that break anyway).
```suggestion
        match byte & 0b11 {
            0b00 => Terrain::Plain,
            0b01 | 0b11 => Terrain::Wall,
            0b10 => Terrain::Swamp,
            // Should be optimized out
            _ => unreachable!("all combinations of 2 bits are covered"),
        }
```